pkg/rpc/server/server.go

```diff
@@ -2036,64 +2022,118 @@ func (s *Server) runScriptInVM(t trigger.Type, script []byte, contractScriptHash
-						Transaction:        tx,
-						NextBlockHeight:    ic.Block.Index,
-					}
-					// Call finalizer manually if MPT-based iterator sessions are enabled. If disabled, then register finalizator.
-					if finalize != nil {
-						finalize()
-						sess.finalize = nil
-					}
-					item = nil
-				}
+				return
 			}
+			sess.iteratorsLock.Lock()
+			if sess.finalize != nil {
+				sess.finalize()
+			}
+			delete(s.sessions, sessionID)
+			sess.iteratorsLock.Unlock()
+		})
+		s.sessionsLock.Lock()
+		if len(s.sessions) >= s.config.SessionPoolSize {
+			s.sessionsLock.Unlock()
+			return nil, response.NewInternalServerError("max session capacity reached")
+		}
+		s.sessions[sessionID] = sess
+		s.sessionsLock.Unlock()
+	} else {
+		ic.Finalize()
+	}
+	var diag *result.InvokeDiag
+	tree := ic.VM.GetInvocationTree()
+	if tree != nil {
+		diag = &result.InvokeDiag{
+			Invocations: tree.Calls,
+			Changes:     storage.BatchToOperations(ic.DAO.GetBatch()),
+		}
+	}
+	notifications := ic.Notifications
+	if notifications == nil {
+		notifications = make([]state.NotificationEvent, 0)
+	}
+	res := &result.Invoke{
+		State:          ic.VM.State().String(),
+		GasConsumed:    ic.VM.GasConsumed(),
+		Script:         script,
+		Stack:          items,
+		FaultException: faultException,
+		Notifications:  notifications,
+		Diagnostics:    diag,
+		Session:        id,
+	}
+
+	return res, nil
+}
+
+// postProcessExecStack changes iterator interop items according to the server configuration.
+// It does modifications in-place, but it returns a session if any iterator was registered.
+func (s *Server) postProcessExecStack(stack []stackitem.Item) *session {
+	var sess session
+
+	for i, v := range stack {
+		var id uuid.UUID
+
+		stack[i], id = s.registerOrDumpIterator(v)
+		if id != (uuid.UUID{}) {
 			sess.iteratorIdentifiers = append(sess.iteratorIdentifiers, &iteratorIdentifier{
-				ID:         iterID.String(),
-				Item:       item,
-				StackIndex: stackIndex,
+				ID:   id.String(),
+				Item: v,
 			})
-			s.sessions[sessionID] = sess
-			s.sessionsLock.Unlock()
-			return iterID, nil
 		}
 	}
-	return result.NewInvoke(ic, script, faultException, registerIterator, s.config.MaxIteratorResultItems), nil
+	if len(sess.iteratorIdentifiers) != 0 {
+		return &sess
+	}
+	return nil
+}
+
+// registerOrDumpIterator changes iterator interop stack items into result.Iterator
+// interop stack items and returns a uuid for it if sessions are enabled. All the other stack
+// items are not changed.
+func (s *Server) registerOrDumpIterator(item stackitem.Item) (stackitem.Item, uuid.UUID) {
+	var iterID uuid.UUID
+
+	if (item.Type() != stackitem.InteropT) || !iterator.IsIterator(item) {
+		return item, iterID
+	}
+	var resIterator result.Iterator
+
+	if s.config.SessionEnabled {
+		iterID = uuid.New()
+		resIterator.ID = &iterID
+	} else {
+		resIterator.Values, resIterator.Truncated = iterator.ValuesTruncated(item, s.config.MaxIteratorResultItems)
```

Comment:
I like the way it's implemented now, but just for the record: it differs from the way C# node works, see https://github.com/neo-project/neo-modules/pull/715#discussion_r897635424. C# node puts iterator ID to the resulting JSON even if sessions are disabled.

Reply:
That's the way it is since #2555, nothing changed there. Disabled sessions make this ID absolutely useless, so we better keep the old behavior of (partial) iterator expansion.